Vectorize  {Last}IndexOfAny{Except} for ASCII needles

MihaZupan opened the pull request:
Contributes to #68328 (and contributes the workhorse implementation for an eventual dedicated API)
cc: @gfoidl

This PR adds a vectorized path for `IndexOfAny`-like methods if the input is at least 8 characters long and the needle is only ASCII.
It is used if `Ssse3` or `AdvSimd.Arm64` are supported.

I also unified all `(Last)IndexOfAny(Except)` methods to use the same approach of
```c#
if (HaystackIsShort) return SimpleSearch();
if (NeedleIsAscii) return Vectorized();
return ProbabilisticMap();
```
leading to perf improvements for both short and long inputs.

For inputs with long runs of no matches, the vectorized path has 10 to 140+ times the throughput on realistic inputs.
As the `-Except` methods were previously `O(n * m)`, you can of course see arbitrarily large improvements. 

<details>
<summary><strong>X86 numbers</strong></summary>

These numbers were collected on a Windows Azure VM running on an Intel Xeon 8370C processor.

|               Method | Toolchain | Length |       Needle |         Mean |     Error | Ratio |
|--------------------- |---------- |------- |------------- |-------------:|----------:|------:|
|           IndexOfAny |      main |      1 |       ABCDEF |     9.787 ns | 0.0087 ns |  1.00 |
|           IndexOfAny |        pr |      1 |       ABCDEF |     5.852 ns | 0.0053 ns |  0.60 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |      1 |       ABCDEF |     5.313 ns | 0.0077 ns |  1.00 |
|     IndexOfAnyExcept |        pr |      1 |       ABCDEF |     6.676 ns | 0.0104 ns |  1.26 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |      1 |       ABCDEF |    11.376 ns | 0.0065 ns |  1.00 |
|       LastIndexOfAny |        pr |      1 |       ABCDEF |     5.783 ns | 0.0037 ns |  0.51 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |      1 |       ABCDEF |     5.946 ns | 0.0123 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |      1 |       ABCDEF |     6.030 ns | 0.0324 ns |  1.01 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |      1 | AlphaNumeric |    63.653 ns | 0.0265 ns |  1.00 |
|           IndexOfAny |        pr |      1 | AlphaNumeric |     5.855 ns | 0.0066 ns |  0.09 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |      1 | AlphaNumeric |     6.371 ns | 0.0128 ns |  1.00 |
|     IndexOfAnyExcept |        pr |      1 | AlphaNumeric |     7.619 ns | 0.0104 ns |  1.20 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |      1 | AlphaNumeric |    74.164 ns | 0.0712 ns |  1.00 |
|       LastIndexOfAny |        pr |      1 | AlphaNumeric |     5.558 ns | 0.0011 ns |  0.07 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |      1 | AlphaNumeric |     8.766 ns | 0.2798 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |      1 | AlphaNumeric |     6.859 ns | 0.0018 ns |  0.88 |
|                      |           |        |              |              |           |       |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |      8 |       ABCDEF |     14.12 ns |  0.016 ns |  1.00 |
|           IndexOfAny |        pr |      8 |       ABCDEF |     12.25 ns |  0.006 ns |  0.87 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |      8 |       ABCDEF |     24.38 ns |  0.019 ns |  1.00 |
|     IndexOfAnyExcept |        pr |      8 |       ABCDEF |     14.86 ns |  0.005 ns |  0.61 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |      8 |       ABCDEF |     16.51 ns |  0.008 ns |  1.00 |
|       LastIndexOfAny |        pr |      8 |       ABCDEF |     12.23 ns |  0.006 ns |  0.74 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |      8 |       ABCDEF |     23.42 ns |  0.022 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |      8 |       ABCDEF |     14.93 ns |  0.013 ns |  0.64 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |      8 | AlphaNumeric |     67.53 ns |  0.075 ns |  1.00 |
|           IndexOfAny |        pr |      8 | AlphaNumeric |     34.92 ns |  0.045 ns |  0.52 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |      8 | AlphaNumeric |     27.99 ns |  0.025 ns |  1.00 |
|     IndexOfAnyExcept |        pr |      8 | AlphaNumeric |     31.09 ns |  0.036 ns |  1.11 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |      8 | AlphaNumeric |     79.39 ns |  0.065 ns |  1.00 |
|       LastIndexOfAny |        pr |      8 | AlphaNumeric |     34.32 ns |  0.028 ns |  0.43 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |      8 | AlphaNumeric |     27.67 ns |  0.097 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |      8 | AlphaNumeric |     25.80 ns |  0.014 ns |  0.93 |
|                      |           |        |              |              |           |       |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |     16 |       ABCDEF |     20.15 ns |  0.060 ns |  1.00 |
|           IndexOfAny |        pr |     16 |       ABCDEF |     12.25 ns |  0.007 ns |  0.61 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |     16 |       ABCDEF |     43.59 ns |  0.043 ns |  1.00 |
|     IndexOfAnyExcept |        pr |     16 |       ABCDEF |     14.96 ns |  0.011 ns |  0.34 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |     16 |       ABCDEF |     21.76 ns |  0.012 ns |  1.00 |
|       LastIndexOfAny |        pr |     16 |       ABCDEF |     12.23 ns |  0.006 ns |  0.56 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |     16 |       ABCDEF |     41.38 ns |  0.033 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |     16 |       ABCDEF |     14.92 ns |  0.010 ns |  0.36 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |     16 | AlphaNumeric |     73.53 ns |  0.060 ns |  1.00 |
|           IndexOfAny |        pr |     16 | AlphaNumeric |     67.08 ns |  0.084 ns |  0.91 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |     16 | AlphaNumeric |     49.79 ns |  0.032 ns |  1.00 |
|     IndexOfAnyExcept |        pr |     16 | AlphaNumeric |     54.95 ns |  0.051 ns |  1.10 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |     16 | AlphaNumeric |     84.76 ns |  0.061 ns |  1.00 |
|       LastIndexOfAny |        pr |     16 | AlphaNumeric |     66.56 ns |  0.060 ns |  0.79 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |     16 | AlphaNumeric |     50.64 ns |  0.162 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |     16 | AlphaNumeric |     45.80 ns |  0.062 ns |  0.91 |
|                      |           |        |              |              |           |       |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |     32 |       ABCDEF |     29.04 ns |  0.034 ns |  1.00 |
|           IndexOfAny |        pr |     32 |       ABCDEF |     13.37 ns |  0.009 ns |  0.46 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |     32 |       ABCDEF |     83.02 ns |  0.093 ns |  1.00 |
|     IndexOfAnyExcept |        pr |     32 |       ABCDEF |     15.98 ns |  0.007 ns |  0.19 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |     32 |       ABCDEF |     31.55 ns |  0.022 ns |  1.00 |
|       LastIndexOfAny |        pr |     32 |       ABCDEF |     13.92 ns |  0.050 ns |  0.44 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |     32 |       ABCDEF |     82.09 ns |  0.090 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |     32 |       ABCDEF |     15.77 ns |  0.007 ns |  0.19 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |     32 | AlphaNumeric |     83.50 ns |  0.061 ns |  1.00 |
|           IndexOfAny |        pr |     32 | AlphaNumeric |     65.77 ns |  0.252 ns |  0.79 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |     32 | AlphaNumeric |    106.94 ns |  0.054 ns |  1.00 |
|     IndexOfAnyExcept |        pr |     32 | AlphaNumeric |     66.03 ns |  0.099 ns |  0.62 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |     32 | AlphaNumeric |     94.24 ns |  0.052 ns |  1.00 |
|       LastIndexOfAny |        pr |     32 | AlphaNumeric |     64.81 ns |  0.073 ns |  0.69 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |     32 | AlphaNumeric |    107.73 ns |  0.249 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |     32 | AlphaNumeric |     65.58 ns |  0.034 ns |  0.61 |
|                      |           |        |              |              |           |       |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |     64 |       ABCDEF |     51.95 ns |  0.321 ns |  1.00 |
|           IndexOfAny |        pr |     64 |       ABCDEF |     15.31 ns |  0.009 ns |  0.30 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |     64 |       ABCDEF |    165.25 ns |  0.166 ns |  1.00 |
|     IndexOfAnyExcept |        pr |     64 |       ABCDEF |     17.54 ns |  0.026 ns |  0.11 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |     64 |       ABCDEF |     52.16 ns |  0.039 ns |  1.00 |
|       LastIndexOfAny |        pr |     64 |       ABCDEF |     15.83 ns |  0.020 ns |  0.30 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |     64 |       ABCDEF |    164.63 ns |  0.167 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |     64 |       ABCDEF |     17.01 ns |  0.008 ns |  0.10 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |     64 | AlphaNumeric |    105.16 ns |  0.079 ns |  1.00 |
|           IndexOfAny |        pr |     64 | AlphaNumeric |     67.26 ns |  0.093 ns |  0.64 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |     64 | AlphaNumeric |    208.43 ns |  0.070 ns |  1.00 |
|     IndexOfAnyExcept |        pr |     64 | AlphaNumeric |     68.18 ns |  0.107 ns |  0.33 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |     64 | AlphaNumeric |    115.14 ns |  0.057 ns |  1.00 |
|       LastIndexOfAny |        pr |     64 | AlphaNumeric |     66.35 ns |  0.040 ns |  0.58 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |     64 | AlphaNumeric |    212.70 ns |  0.583 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |     64 | AlphaNumeric |     67.31 ns |  0.059 ns |  0.32 |
|                      |           |        |              |              |           |       |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |    128 |       ABCDEF |     90.31 ns |  0.083 ns |  1.00 |
|           IndexOfAny |        pr |    128 |       ABCDEF |     18.74 ns |  0.008 ns |  0.21 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |    128 |       ABCDEF |    336.07 ns |  0.234 ns |  1.00 |
|     IndexOfAnyExcept |        pr |    128 |       ABCDEF |     20.78 ns |  0.015 ns |  0.06 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |    128 |       ABCDEF |    100.96 ns |  0.050 ns |  1.00 |
|       LastIndexOfAny |        pr |    128 |       ABCDEF |     18.80 ns |  0.024 ns |  0.19 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |    128 |       ABCDEF |    337.61 ns |  0.233 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |    128 |       ABCDEF |     21.00 ns |  0.040 ns |  0.06 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |    128 | AlphaNumeric |    148.35 ns |  0.231 ns |  1.00 |
|           IndexOfAny |        pr |    128 | AlphaNumeric |     70.20 ns |  0.121 ns |  0.47 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |    128 | AlphaNumeric |    425.60 ns |  0.178 ns |  1.00 |
|     IndexOfAnyExcept |        pr |    128 | AlphaNumeric |     71.25 ns |  0.041 ns |  0.17 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |    128 | AlphaNumeric |    165.64 ns |  0.085 ns |  1.00 |
|       LastIndexOfAny |        pr |    128 | AlphaNumeric |     70.08 ns |  0.118 ns |  0.42 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |    128 | AlphaNumeric |    438.12 ns |  1.407 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |    128 | AlphaNumeric |     72.16 ns |  0.126 ns |  0.16 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |    256 |       ABCDEF |    182.60 ns |  0.382 ns |  1.00 |
|           IndexOfAny |        pr |    256 |       ABCDEF |     29.07 ns |  0.026 ns |  0.16 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |    256 |       ABCDEF |    703.61 ns |  0.468 ns |  1.00 |
|     IndexOfAnyExcept |        pr |    256 |       ABCDEF |     32.34 ns |  0.019 ns |  0.05 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |    256 |       ABCDEF |    196.47 ns |  1.878 ns |  1.00 |
|       LastIndexOfAny |        pr |    256 |       ABCDEF |     29.52 ns |  0.063 ns |  0.15 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |    256 |       ABCDEF |    704.98 ns |  0.586 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |    256 |       ABCDEF |     31.83 ns |  0.019 ns |  0.05 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |    256 | AlphaNumeric |    236.97 ns |  0.128 ns |  1.00 |
|           IndexOfAny |        pr |    256 | AlphaNumeric |     81.93 ns |  0.287 ns |  0.35 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |    256 | AlphaNumeric |    878.10 ns |  0.445 ns |  1.00 |
|     IndexOfAnyExcept |        pr |    256 | AlphaNumeric |     82.96 ns |  0.094 ns |  0.09 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |    256 | AlphaNumeric |    247.84 ns |  0.110 ns |  1.00 |
|       LastIndexOfAny |        pr |    256 | AlphaNumeric |     80.38 ns |  0.055 ns |  0.32 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |    256 | AlphaNumeric |    871.79 ns |  0.353 ns |  1.00 |
| LastIndexOfAnyExcept |        pr |    256 | AlphaNumeric |     83.21 ns |  0.096 ns |  0.10 |
|                      |           |        |              |              |           |       |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |  10000 |       ABCDEF |   6,409.6 ns |   4.42 ns |  1.00 |
|           IndexOfAny |        pr |  10000 |       ABCDEF |     651.2 ns |   0.36 ns |  0.10 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |  10000 |       ABCDEF |  71,960.2 ns |  31.30 ns |  1.00 |
|     IndexOfAnyExcept |        pr |  10000 |       ABCDEF |     741.1 ns |   0.36 ns |  0.01 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |  10000 |       ABCDEF |   6,396.7 ns |   2.27 ns |  1.00 |
|       LastIndexOfAny |        pr |  10000 |       ABCDEF |     643.2 ns |   0.36 ns |  0.10 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |  10000 |       ABCDEF |  81,927.2 ns |  23.11 ns | 1.000 |
| LastIndexOfAnyExcept |        pr |  10000 |       ABCDEF |     728.1 ns |   0.26 ns | 0.009 |
|                      |           |        |              |              |           |       |
|           IndexOfAny |      main |  10000 | AlphaNumeric |   6,450.1 ns |   2.54 ns |  1.00 |
|           IndexOfAny |        pr |  10000 | AlphaNumeric |     704.2 ns |   0.41 ns |  0.11 |
|                      |           |        |              |              |           |       |
|     IndexOfAnyExcept |      main |  10000 | AlphaNumeric | 114,854.9 ns |  91.93 ns | 1.000 |
|     IndexOfAnyExcept |        pr |  10000 | AlphaNumeric |     791.0 ns |   0.36 ns | 0.007 |
|                      |           |        |              |              |           |       |
|       LastIndexOfAny |      main |  10000 | AlphaNumeric |   6,613.7 ns |  21.32 ns |  1.00 |
|       LastIndexOfAny |        pr |  10000 | AlphaNumeric |     690.0 ns |   0.08 ns |  0.10 |
|                      |           |        |              |              |           |       |
| LastIndexOfAnyExcept |      main |  10000 | AlphaNumeric | 110,475.7 ns |  74.60 ns | 1.000 |
| LastIndexOfAnyExcept |        pr |  10000 | AlphaNumeric |     774.9 ns |   0.04 ns | 0.007 |

</details>

On X86 we have to do a bit more work if the needle contains a zero:
|     Method | Length |         Needle |     Mean |   Error |
|----------- |------- |--------------- |---------:|--------:|
| IndexOfAny |  10000 |         ABCDEF | 536.8 ns | 0.45 ns |
| IndexOfAny |  10000 | NeedleWithZero | 820.7 ns | 0.74 ns |

Approximate ARM64 numbers:
|            Method | Length | Needle |      Mean |     Error |
|------------------ |------- |------- |----------:|----------:|
|        IndexOfAny |  10000 | ABCDEF |  1.783 µs | 0.0051 µs |
| CurrentIndexOfAny |  10000 | ABCDEF | 10.355 µs | 0.0010 µs |

stephentoub commented:
Excellent, thanks for working on this.

> The (Last)IndexOfAnyExcept methods that were previously using an O(n * m) loop regress with this change if the needle is long, while the haystack is small

Have we investigated doing something different for short haystacks? e.g. something super simple like:
```C#
if (haystack.Length < Vector128<short>.Count)
{
    for (int i = 0; i < haystack.Length; i++)
        if (needle.Contains(haystack[i]))
            return i;
    return -1;
}
...
```
?

MihaZupan commented:
That is what we do for the `-Except` overloads with this PR.
```c#
haystack.Length >= 8 ? Vectorized() : SimpleLoop()
```
whereas `(Last)IndexOfAny` does
```c#
haystack.Length >= 8 ? Vectorized() : ProbabilisticMap()
```
In places where we were already using the `ProbabilisticMap`, the init overhead seems to be very similar.

It would likely be beneficial to tweak the exact cutoff and add the `SimpleLoop` variant where we currently only use the `ProbabilisticMap`.
E.g. (made up constants)
```c#
int IndexOfAny()
{
    if (haystack.Length < Max(8, needle.Length / 2))
        return SimpleLoop();

    if (IsAscii(needle))
        return Vectorized();

    return ProbabilisticMap();
}

int IndexOfAnyExcept()
{
    if (haystack.Length >= Max(8, needle.Length / 2) && IsAscii(needle))
        return Vectorized();

    return SimpleLoop();
}
```

I can look into what sort of numbers we'd see with something like that, though I would hope that in general, a dedicated API to hide the init cost completely would be the preferred approach.

stephentoub commented:
> though I would hope that in general, a dedicated API to hide the init cost completely would be the preferred approach

Certainly for cases where you're going to be invoking something repeatedly and have the foresight to create and cache the preprocessed vector information.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+        private struct Negate : INegator
+        {
+            public static bool NegateIfNeeded(bool result) => !result;
+            public static Vector128<byte> NegateIfNeeded(Vector128<byte> result) => Vector128.Equals(result, Vector128<byte>.Zero);
```
```suggestion
            public static Vector128<byte> NegateIfNeeded(Vector128<byte> result) => ~result;
```

MihaZupan replied:
These aren't equivalent. I explicitly want to check whether any value was 0.
The name may not be the best though, suggestions are welcome :)

gfoidl replied:
Got you, but I don't have any real idea for a better name. But a comment explaining to check for 0 here is the intention would clarify this.

For names maybe `CompareToZeroIfNeeded`? This doesn't fit to `INegator` anyway. I think as internal implementation detail the comment is the best option -- except someone else comes up with a proper name.

gfoidl reviewed (commented):
👍🏻 
Left some notes.

I expect these APIs should work for byte-inputs too, so with another static abstract interface (probably my favorite C# 11 feature) this can be done without duplication. I'm re-writing the base64 code in my repo using [Char- and Byte-Operations](https://github.com/gfoidl/Base64/tree/293a09bcde0b8a3ee8d596cda288119e976039de/source/gfoidl.Base64/Internal/Operations) for this* -- i.e. reading two char/short-vectors and combining them into one byte vector for further processing. Something similar could be done here too.

\* it's still WIP, so pack signed vs. unsigned isn't done there

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+            // ARM64: Take the low byte of each character.
+            // - All values result in (value & 0xFF).
+            Vector128<byte> source = Sse2.IsSupported
+                ? Sse2.PackUnsignedSaturate(source0, source1)
```
I'm not sure that gives you anything though, just shifts the problem to now having to explicitly handle `127`.

MihaZupan replied:
Might be, but you have to check 0 regardless because of the >= 32768 range where `PackSignedSaturate` will produce values > 127 and Shuffle will spit out 0s again.
Unless I'm missing some other trick?

gfoidl replied:
> \> 127 and Shuffle will spit out 0s again

Ahhh, forgot about this. I'm sure there's some trick for this...I have to work on Sunday on something similar, so if I find something I'll let you know.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+            // - Values  > 32767 result in 0. Because of this we must do more work to handle needles that contain 0.
+            // ARM64: Take the low byte of each character.
+            // - All values result in (value & 0xFF).
+            Vector128<byte> source = Sse2.IsSupported
+                ? Sse2.PackUnsignedSaturate(source0, source1)
+                : AdvSimd.Arm64.UnzipEven(source0.AsByte(), source1.AsByte());
```
Even with such an API, it may not be optimal to use it here if it's any more expensive than `UnzipEven` on ARM, as we don't make use of any saturation side effects later on.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+
+            if (searchSpaceLength > 2 * Vector128<short>.Count)
+            {
+                ref short twoVectorsAwayFromEnd = ref Unsafe.Add(ref searchSpace, searchSpaceLength - (2 * Vector128<short>.Count));
```
It was intentional to not loop again for the last 16 if length was exactly a multiple of 16.
I'll add a comment.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<byte> Shuffle(Vector128<byte> vector, Vector128<byte> indices)
+        {
+            // Vector128.Shuffle is emitting a software fallback as indices are not constant
```
As far as I understood Taner, you are correct, but in this case, these will never be constants.
The vector may be constant while the indices are the user input.

gfoidl replied:
Sure. My suggestion is just about the comment to have it precise -- kind of nitpicking 😉

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+            if (IsSupportedAndProfitable(searchSpaceLength))
+            {
+                Vector128<byte> bitmap = default;
+                if (TryComputeBitmap(asciiValues, (byte*)&bitmap))
+                {
+                    index = Ssse3.IsSupported && (bitmap.GetElementUnsafe(0) & 1) != 0
```
I would expect that to be more expensive with the extra branch for each element in the needle though. Looking at the first bit of the computed vector is super cheap.

gfoidl replied:
Yep, I meant the "Edit"-solution as
* the code here looks nicer
* I think / hope the JIT can do better dead-code elimination here (only gut feeling)

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+                {
+                    index = Ssse3.IsSupported && (bitmap.GetElementUnsafe(0) & 1) != 0
+                        ? IndexOfAnyVectorized<TNegator, Ssse3HandleZeroInNeedle>(ref searchSpace, searchSpaceLength, bitmap)
+                        : IndexOfAnyVectorized<TNegator, Fallback>(ref searchSpace, searchSpaceLength, bitmap);
```
Yeah, I didn't like the name either 😆
If others like `Default` better here I'll change it.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+                Vector128<byte> bitmap = default;
+                if (TryComputeBitmap(asciiValues, (byte*)&bitmap))
+                {
+                    index = Ssse3.IsSupported && (bitmap.GetElementUnsafe(0) & 1) != 0
```
I'll double-check and change it in that case, thanks!

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+        {
+            // Pack two vectors of characters into bytes.
+            // X86: Downcast every character using saturation.
+            // - Values <= 32767 result in min(value, 255).
+            // - Values  > 32767 result in 0. Because of this we must do more work to handle needles that contain 0.
```
I'll clarify that in the comment. I'm really thinking of these values as `ushort`s (chars), using `short` is just more convenient to avoid more `AsInt16` calls.

gfoidl commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,333 @@
+                int highNibble = c >> 4;
+                int lowNibble = c & 0xF;
+
+                bitmapLocal[lowNibble] |= (byte)(1 << highNibble);
```
```suggestion
                bitmapLocal[(uint)lowNibble] |= (byte)(1 << highNibble);
```
To get rid of the sign-extending move.

MihaZupan commented:
Rerunning the benchmarks with https://github.com/MihaZupan/runtime/commit/e06c007df6552892f642c2a8d1bdafb60be3892b now.

MihaZupan commented:
I added a fast path for short haystacks to the probabilistic code path, leading to nice improvements there (I updated the numbers in the top post). E.g.
|               Method | Toolchain | Length |       Needle |         Mean |     Error | Ratio |
|--------------------- |---------- |------- |------------- |-------------:|----------:|------:|
|       LastIndexOfAny |      main |      1 | AlphaNumeric |    74.164 ns | 0.0712 ns |  1.00 |
|       LastIndexOfAny |        pr |      1 | AlphaNumeric |     5.558 ns | 0.0011 ns |  0.07 |

The only real regression is the `(Last)IndexOfAnyExcept` for long needles and short haystacks as mentioned before. I'll keep playing around with cutoffs to try and improve that case too.

MihaZupan commented:
The cutoff for the simple loop is now:
```c#
searchSpaceLength < Vector128<short>.Count || (searchSpaceLength < 20 && searchSpaceLength < (valuesLength >> 1))
```

Updated the benchmarks above, this is now a win pretty much across the board.

MihaZupan commented:
/azp run runtime

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,345 @@
+    {
+        private static bool IsSupported => Ssse3.IsSupported || AdvSimd.Arm64.IsSupported;
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
There appears to be AggressiveInlining all the way up the stack.  Is this going to end up inlining everything into the call site and bloating every use of the IndexOfAny overload?  Will it result in hitting up against any JIT limits that cause other things which should be inlined to not be?

MihaZupan replied:
The init logic is inlined into the `ProbabilisticMap` methods, but those shouldn't inline further up into `MemoryExtensions`.
That is, these are intentionally not marked with `AggressiveInlining`:
https://github.com/dotnet/runtime/blob/61bc1863ce8ccbd64c4d7949b259b338918d4326/src/libraries/System.Private.CoreLib/src/System/ProbabilisticMap.cs#L79-L89

MihaZupan replied:
Let me double-check. I assume marking these with `NoInlining` should be sufficient?

stephentoub replied:
> I assume marking these with NoInlining should be sufficient?

Yes, but @EgorBo to confirm, and to comment on whether there are any unforeseen downsides to that.

Alternatively, we could be more sparing in our use of AggressiveInlining.

MihaZupan replied:
Ah, it is being inlined 😄

In the case of `(Last)IndexOfAny`, all the way into the benchmark method.
In the case of `(Last)IndexOfAnyExcept`, into the `MemoryExtensions` helper.

This may also have been the cause of some other weirdness I saw on some machines (having to split `ProbabilisticIndexOfAny` into a separate method). I'll get this fixed.

On a related note, is there a reason the `-Except` methods aren't marked with `AggressiveInlining` whereas regular `IndexOfAny` methods are?

stephentoub replied:
> On a related note, is there a reason the -Except methods aren't marked with AggressiveInlining whereas regular IndexOfAny methods are?

Which are you referring to?

MihaZupan replied:
https://github.com/dotnet/runtime/blob/6811cc4dcb4aaabbfa471c0bff101ee70b237fc0/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L1297-L1298
https://github.com/dotnet/runtime/blob/6811cc4dcb4aaabbfa471c0bff101ee70b237fc0/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L1566-L1567

vs.
https://github.com/dotnet/runtime/blob/6811cc4dcb4aaabbfa471c0bff101ee70b237fc0/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L681-L682
https://github.com/dotnet/runtime/blob/6811cc4dcb4aaabbfa471c0bff101ee70b237fc0/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L924-L925

stephentoub replied:
I don't know, but a) we should make these consistent, and b) we should seriously reconsider the aggressive inlining that's on there (or consider how IsKnownConstant could be used here).  If the values passed is a constant, then the JIT can do a good job of trimming away all of these branches.  But if it's not, it results in significant code bloat to the call site, e.g. this:
```C#
private static bool Contains(string text, string values) =>
    text.AsSpan().IndexOfAny(values) >= 0;
```
results in this:
```
; Assembly listing for method ConsoleApp15.Program:Contains(System.String,System.String):bool
; Emitting BLENDED_CODE for X64 CPU with AVX - Windows
; optimized code
; rsp based frame
; partially interruptible
; No PGO data
; 2 inlinees with PGO data; 23 single block inlinees; 1 inlinees without PGO data

G_M000_IG01:                ;; offset=0000H
       4883EC48             sub      rsp, 72
 
G_M000_IG02:                ;; offset=0004H
       4885C9               test     rcx, rcx
       7475                 je       SHORT G_M000_IG10
 
G_M000_IG03:                ;; offset=0009H
       4C8D410C             lea      r8, bword ptr [rcx+0CH]
       8B4908               mov      ecx, dword ptr [rcx+08H]
 
G_M000_IG04:                ;; offset=0010H
       8BC1                 mov      eax, ecx
       4885D2               test     rdx, rdx
       7442                 je       SHORT G_M000_IG08
 
G_M000_IG05:                ;; offset=0017H
       488D4A0C             lea      rcx, bword ptr [rdx+0CH]
       448B4A08             mov      r9d, dword ptr [rdx+08H]
 
G_M000_IG06:                ;; offset=001FH
       8BD0                 mov      edx, eax
       89542444             mov      dword ptr [rsp+44H], edx
       48894C2438           mov      bword ptr [rsp+38H], rcx
       4183F905             cmp      r9d, 5
       0F871B010000         ja       G_M000_IG15
       418BC1               mov      eax, r9d
       4C8D0D2A010000       lea      r9, [reloc @RWD00]
       458B0C81             mov      r9d, dword ptr [r9+4*rax]
       4C8D15BBFFFFFF       lea      r10, G_M000_IG02
       4D03CA               add      r9, r10
       41FFE1               jmp      r9
 
G_M000_IG07:                ;; offset=004FH
       B8FFFFFFFF           mov      eax, -1
       E904010000           jmp      G_M000_IG16
 
G_M000_IG08:                ;; offset=0059H
       33C9                 xor      rcx, rcx
       4533C9               xor      r9d, r9d
       EBBF                 jmp      SHORT G_M000_IG06
 
G_M000_IG09:                ;; offset=0060H
       488B4C2438           mov      rcx, bword ptr [rsp+38H]
       480FBF01             movsx    rax, word  ptr [rcx]
       498BC8               mov      rcx, r8
       8BD0                 mov      edx, eax
       448B442444           mov      r8d, dword ptr [rsp+44H]
       FF1587A32000         call     [System.SpanHelpers:IndexOfValueType[short,DontNegate`1[short]](byref,short,int):int]
       E9DF000000           jmp      G_M000_IG16
 
G_M000_IG10:                ;; offset=007EH
       33C9                 xor      rcx, rcx
       4533C0               xor      r8d, r8d
       4987C8               xchg     rcx, r8
       EB88                 jmp      SHORT G_M000_IG04
 
G_M000_IG11:                ;; offset=0088H
       488B4C2438           mov      rcx, bword ptr [rsp+38H]
       4C0FBF09             movsx    r9, word  ptr [rcx]
       480FBF4102           movsx    rax, word  ptr [rcx+02H]
       498BC8               mov      rcx, r8
       418BD1               mov      edx, r9d
       448BC0               mov      r8d, eax
       448B4C2444           mov      r9d, dword ptr [rsp+44H]
       FF15BEBE2000         call     [System.SpanHelpers:IndexOfAnyValueType[short,DontNegate`1[short]](byref,short,short,int):int]
       E9AE000000           jmp      G_M000_IG16
 
G_M000_IG12:                ;; offset=00AFH
       488B4C2438           mov      rcx, bword ptr [rsp+38H]
       4C0FBF09             movsx    r9, word  ptr [rcx]
       480FBF4102           movsx    rax, word  ptr [rcx+02H]
       4C0FBF5104           movsx    r10, word  ptr [rcx+04H]
       89542420             mov      dword ptr [rsp+20H], edx
       498BC8               mov      rcx, r8
       418BD1               mov      edx, r9d
       448BC0               mov      r8d, eax
       458BCA               mov      r9d, r10d
       FF1570C62000         call     [System.SpanHelpers:IndexOfAnyValueType[short,DontNegate`1[short]](byref,short,short,short,int):int]
       E980000000           jmp      G_M000_IG16
 
G_M000_IG13:                ;; offset=00DDH
       488B4C2438           mov      rcx, bword ptr [rsp+38H]
       4C0FBF09             movsx    r9, word  ptr [rcx]
       480FBF4102           movsx    rax, word  ptr [rcx+02H]
       4C0FBF5104           movsx    r10, word  ptr [rcx+04H]
       480FBF4906           movsx    rcx, word  ptr [rcx+06H]
       894C2420             mov      dword ptr [rsp+20H], ecx
       89542428             mov      dword ptr [rsp+28H], edx
       498BC8               mov      rcx, r8
       418BD1               mov      edx, r9d
       448BC0               mov      r8d, eax
       458BCA               mov      r9d, r10d
       FF1559DE2000         call     [System.SpanHelpers:IndexOfAnyValueType[short,DontNegate`1[short]](byref,short,short,short,short,int):int]
       EB4C                 jmp      SHORT G_M000_IG16
 
G_M000_IG14:                ;; offset=0111H
       488B4C2438           mov      rcx, bword ptr [rsp+38H]
       4C0FBF09             movsx    r9, word  ptr [rcx]
       480FBF4102           movsx    rax, word  ptr [rcx+02H]
       4C0FBF5104           movsx    r10, word  ptr [rcx+04H]
       4C0FBF5906           movsx    r11, word  ptr [rcx+06H]
       480FBF4908           movsx    rcx, word  ptr [rcx+08H]
       44895C2420           mov      dword ptr [rsp+20H], r11d
       894C2428             mov      dword ptr [rsp+28H], ecx
       89542430             mov      dword ptr [rsp+30H], edx
       498BC8               mov      rcx, r8
       418BD1               mov      edx, r9d
       448BC0               mov      r8d, eax
       458BCA               mov      r9d, r10d
       FF154B842000         call     [System.SpanHelpers:IndexOfAnyValueType[short](byref,short,short,short,short,short,int):int]
       EB0E                 jmp      SHORT G_M000_IG16
 
G_M000_IG15:                ;; offset=014FH
       498BC8               mov      rcx, r8
       4C8B442438           mov      r8, bword ptr [rsp+38H]
       FF15A3222000         call     [System.MemoryExtensions:IndexOfAnyProbabilistic(byref,int,byref,int):int]
 
G_M000_IG16:                ;; offset=015DH
       F7D0                 not      eax
       C1E81F               shr      eax, 31
 
G_M000_IG17:                ;; offset=0162H
       4883C448             add      rsp, 72
       C3                   ret      
 
RWD00  	dd	0000004Bh ; case G_M000_IG07
       	dd	0000005Ch ; case G_M000_IG09
       	dd	00000084h ; case G_M000_IG11
       	dd	000000ABh ; case G_M000_IG12
       	dd	000000D9h ; case G_M000_IG13
       	dd	0000010Dh ; case G_M000_IG14

; Total bytes of code 359
```

MihaZupan replied:
Opened https://github.com/dotnet/runtime/issues/76843 to track this.

I've removed a bunch of `AggressiveInlining` hints from this PR now. I'll update the top post when I rerun the benchmarks.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,345 @@
+
+            // We have 1-16 characters remaining. Process the first and last vector in the search space.
+            // They may overlap, but we'll handle that in the index calculation if we do get a match.
+            {
```
This logic appears to depend on there being at least a vector's worth of data.  I assume that's being checked by the caller, but it'd be good to assert it.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,345 @@
+
+            if (searchSpaceLength > 2 * Vector128<short>.Count)
+            {
+                // Process the input in chunks of 16 characters. If the input length is a multiple of 16, don't consume
```
We're mainly interested in a single byte from each character, and the cost to combine the two vectors via `PackUnsignedSaturate` is small compared to the rest of the lookup logic. This allows us to double the throughput "for free" by making use of the whole `Vector128` in the core of the lookup.
I'll add a comment.

stephentoub replied:
> I'll add a comment.

Thanks.  It's just not obvious here at the call site; you need to read into the callee to understand that it's basically narrowing two vectors into one.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:
```diff
@@ -702,6 +701,15 @@ public static ReadOnlyMemory<char> AsMemory(this string? text, Range range)
                     return IndexOfAnyExcept(span, values[0], values[1], values[2], values[3]);
 
                 default:
+                    if (RuntimeHelpers.IsBitwiseEquatable<T>() && Unsafe.SizeOf<T>() == sizeof(short))
```
Nit: since we're unsafe casting to char, should we compare against sizeof(char), just for consistency?  Same below.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,345 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static Vector128<byte> Shuffle(Vector128<byte> vector, Vector128<byte> indices)
+        {
+            // Vector128.Shuffle is emitting a software fallback as indices are not given as inline constants.
```
Is there an issue for this? If Sse3.Shuffle can do the "right thing", presumably Vector128.Shuffle should be able to do the right thing if Ssse33.IsSupported.

MihaZupan replied:
This one is about Vector256, but I believe it's the same problem: https://github.com/dotnet/runtime/issues/72793

@tannergooding `Vector128.Shuffle` may emit extra code here to match the behavior across different platforms (not just `Ssse3.Shuffle` and `AdvSimd.Arm64.VectorTableLookup` directly), right?
If that's the case, I can remove the comment here as we wouldn't actually want to use `Vector128.Shuffle`, as we don't need exactly the same behavior.

danmoseley commented:
Curious to see what this does to regex redux 

https://benchmarksgame-team.pages.debian.net/benchmarksgame/program/regexredux-csharpcore-5.html

stephentoub commented:
> Curious to see what this does to regex redux

It won't do anything.  We don't use this API from RegexCompiler / source generator, nor does regex redux have any sets that would trigger it.  We _will_ update RegexCompiler / source generator to use whatever new API we create that plugs into the same implementation this is adding but that let's us precompute the vector/bit set rather than having to do it on each call.

MihaZupan commented:
/azp run runtime

MihaZupan committed:
Vectorize {Last}IndexOfAny{Except} for ASCII needles

MihaZupan committed:
Add a ShouldUseSimpleLoop helper

stephentoub commented on src/libraries/System.Memory/tests/Span/IndexOfAny.char.cs:
```diff
@@ -772,5 +773,70 @@ public static void MakeSureNoChecksGoOutOfRangeMany_Char()
+
+            Span<char> needleSpace = stackalloc char[8];
+            Span<char> haystackSpace = stackalloc char[40];
+            var rng = new Random();
```
To make this reproducable, we should include a seed here.

MihaZupan replied:
If I remember correctly, I removed it because the RNG part represents a significant portion of the test execution time, with the non-explicit-seed ctor using the faster impl AFAIK.

For reproducibility, I made sure to emit the exact inputs that failed. Does that address your concerns here?

stephentoub replied:
> Does that address your concerns here?

Only partially. There are two concerns:
1. That a test could fail and you not know why.
2. That two runs of the same test suite might be non-deterministic.

Outputting the exact inputs addresses (1) but not (2). And (2) in general contributes to test flakiness.

If this test is about randomly stressing the implementation, then I don't believe it belongs in the functional test suite.  If it's about using pseudo-randomness to try out thousands of inputs without having to manually code them all and be creative enough to ensure a reasonable spread of inputs, then it should be done in a way that runs the same tests every time, which means using a seed.

MihaZupan replied:
I changed it to waste less time in Random and use a constant seed.

It was extremely useful when working on the initial implementation to get rid of all the edge-case bugs. Some were so odd that I don't believe we would have come up with test cases for otherwise.

I increased the number of iterations so it now takes about a second of CPU time to run through them, so I moved these to outerloop.
Outerloop already takes ~4 minutes on my machine so I'm gonna assume that's acceptable.

stephentoub commented on src/libraries/System.Memory/tests/Span/IndexOfAny.char.cs:
```diff
@@ -772,5 +773,70 @@ public static void MakeSureNoChecksGoOutOfRangeMany_Char()
+            Span<char> haystackSpace = stackalloc char[40];
+            var rng = new Random();
+
+            for (int iterations = 0; iterations < 100_000; iterations++)
```
How long do these tests take to run?  Wondering if they meaningfully increase execution time of the test suite and should be made outerloop.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,353 @@
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static unsafe bool TryComputeBitmap(ReadOnlySpan<char> values, byte* bitmap, out bool needleContainsZero)
+        {
+            byte* bitmapLocal = bitmap;
```
Why is the local copy needed?  I seem to remember there being an issue about code gen? If so, can we include a link to the open issue?

MihaZupan replied:
I copied this approach from
https://github.com/dotnet/runtime/blob/fbc20578253e0deb7267eea5a02b680c3cfe0e2a/src/libraries/System.Private.CoreLib/src/System/ProbabilisticMap.cs#L40

I'll add a comment here as well

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/IndexOfAnyAsciiSearcher.cs:
```diff
@@ -0,0 +1,353 @@
+            // We want to know if any character didn't match, as that means it should be treated as a match for the -Except method.
+            public static Vector128<byte> NegateIfNeeded(Vector128<byte> result) => Vector128.Equals(result, Vector128<byte>.Zero);
+            public static uint ExtractMask(Vector128<byte> result) => result.ExtractMostSignificantBits();
+        }
```
Don't we already have some of these defined in SpanHelpers or MemoryExtensions?  Is it possible to reuse and/or consolidate the types and associated helpers?

MihaZupan replied:
We have structs named like these, but the ones in this file rely heavily on the implementation details - e.g. `ExtractMask` does more work than strictly extracting the mask in `DontNegate`, or `NegateIfNeeded` does something slightly different than strictly just negating in `Negate`.

The set of helpers in `ProbabilisticMap.cs` can probably delegate to the one in `MemoryExtensions` though.

MihaZupan replied:
Changed the ones in `ProbabilisticMap` to reuse the ones in `SpanHelpers`.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs:
```diff
@@ -1791,8 +1779,9 @@ private static unsafe int IndexOfAnyProbabilistic(ref char searchSpace, int sear
         /// <param name="values">The set of values to search for.</param>
-        public static int LastIndexOfAny<T>(this Span<T> span, ReadOnlySpan<T> values) where T : IEquatable<T>?
-            => LastIndexOfAny((ReadOnlySpan<T>)span, values);
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
This was necessary?  Or just added for consistency with something?

MihaZupan replied:
This was removed by mistake in https://github.com/dotnet/runtime/pull/75754#discussion_r1005797894

I added it back to at least make it consistent with the rest of `{Last}IndexOfAny` overloads.

stephentoub commented:
> Updated the benchmarks above, this is now a win pretty much across the board.

Meaning with this PR, for long or short inputs, vectorized or not, ASCII or not, everything you've tried is as good or better than it was before?  If so, yay!

MihaZupan commented:
> Meaning with this PR, for long or short inputs, vectorized or not, ASCII or not, everything you've tried is as good or better than it was before? If so, yay!

Yes.

The main scenario I can think of that would regress is if the needle contains non-ASCII, but starts with ASCII (e.g. `alphabet + 'ü'`) as you now have to pay for an extra scan to build the ASCII bitmap that you don't actually need.

MihaZupan committed:
Review feedback

MihaZupan commented:
/azp run runtime-libraries-coreclr outerloop

MihaZupan commented:
Test failures look like #76755, #77726
